ichnaea/service/locate.py

```diff
@@ -122,496 +100,536 @@ def map_data(data):
-
-        if geoip['country_code'] not in cell_countries:
-            if cell_countries:
-                stats_client.incr('%s.anomaly.geoip_mcc_mismatch' % api_name)
-            # Only use the GeoIP country as an additional possible match,
-            # but retain the cell countries as a likely match as well.
-            cell_countries.append(geoip['country_code'])
-
-        stats_client.incr('%s.country_from_geoip' % api_name)
-        geoip_res = {
-            'lat': geoip['latitude'],
-            'lon': geoip['longitude'],
-            'accuracy': geoip['accuracy'],
+    def prepare_location(self, queried_objects):
+        self.stat_count('cell_lac_found')
+        self.stat_count('cell_lac_hit')
+        # take the smallest LAC of any the user is inside
+        lac = sorted(queried_objects, key=operator.attrgetter('range'))[0]
+        accuracy = max(LAC_MIN_ACCURACY, lac.range)
+        accuracy = float(accuracy)
+        return {
+            'lat': lac.lat,
+            'lon': lac.lon,
+            'accuracy': accuracy,
         }
-        return (geoip_res, most_common_elements(cell_countries))
 
-    else:
-        stats_client.incr('%s.no_geoip_found' % api_name)
-
-    # Pick the most-commonly-occurring country codes if we got any
-    cc = most_common_elements(cell_countries)
-    if cc:
-        stats_client.incr('%s.country_from_mcc' % api_name)
-        return (None, cc)
-
-    stats_client.incr('%s.no_country' % api_name)
-    return (None, [])
-
-
-def search_cell(session, cells, stats_client, api_name):
-    if not cells:
-        return
-
-    length = len(cells)
-    avg_lat = sum([c.lat for c in cells]) / length
-    avg_lon = sum([c.lon for c in cells]) / length
-    return {
-        'lat': avg_lat,
-        'lon': avg_lon,
-        'accuracy': estimate_accuracy(avg_lat, avg_lon,
-                                      cells, CELL_MIN_ACCURACY),
-    }
 
+class WifiLocationProvider(AbstractLocationProvider):
+    log_name = 'wifi'
 
-def search_cell_lac(session, lacs, stats_client, api_name):
-    if not lacs:
-        return
+    def cluster_elements(self, items, distance_fn, threshold):
+        """
+        Generic pairwise clustering routine.
 
-    # take the smallest LAC of any the user is inside
-    lac = sorted(lacs, key=operator.attrgetter('range'))[0]
-    accuracy = max(LAC_MIN_ACCURACY, lac.range)
-    accuracy = float(accuracy)
-    return {
-        'lat': lac.lat,
-        'lon': lac.lon,
-        'accuracy': accuracy,
-    }
+        :param items: A list of elemenets to cluster.
+        :param distance_fn: A pairwise distance_fnance function over elements.
+        :param threshold: A numeric thresholdold for clustering;
+                          clusters P, Q will be joined if
+                          distance_fn(a,b) <= threshold,
+                          for any a in P, b in Q.
 
+        :returns: A list of lists of elements, each sub-list being a cluster.
+        """
```

Comment:
In order for the doc string to be picked up, `docs/internal_api/locate.rst` needs to change from an autofunction to a autoclass (http://sphinx-doc.org/ext/autodoc.html#directive-autoclass).